### PR TITLE
64 simple mpi support

### PR DESCRIPTION
**Pull request by hiker**
This PR adds very simple (and not particularly efficient) MPI support:
1) The constructor of fields takes an optional argument which is the global array (on each PE). Each PE then copies the data from its own subdomain into the field. That requires that the global array is created on each PE (instead of doing it on one PE, and then broadcasting the data), but it works really well with my example for the PSyclone training. The advantage of providing the global array on each PE is that there is no need during input handling to test ``if (on_master())``, it is simply:
```
    allocate(initial_state(n_cols, n_rows))

    call get_initial_state(initial_state, n_rows, n_cols)

    initial = r2d_field(grid, grid_points = GO_T_POINTS,   &
                        init_global_data = initial_state)
```
2) There is a new function to 'gather' a distributed array back into a 2d field. The output function still needs a `if(on_master())`, (so that only one process outputs the results)

### CODE REVIEW

**Comment by hiker** [comment]
Maybe an example how output now works:
```
        call field%gather_inner_data(global_data)

        if (on_master()) then
            do j=lbound(global_data, 2), ubound(global_data, 2)
                write(*,"(99F2.0)") global_data(1:ubound(global_data,1), j)
            enddo
            write(*,*)
        endif
        deallocate(global_data)
```

**Comment by hiker** [comment]
Ready for a first review. I am happy to discuss if this approach should be too limited (admittedly it was tweaked to simplify the PSyclone training).

**Comment by hiker** [comment]
@arporter Hi Andy, would it be possible to get this reviewed in the near future? I am working on the PSyclone training, and atm I have to point to an external dl_esm_info repo (instead of the internal PSyclone one), and have to take care when compiling any psydata lib to also point to the same external repo (otherwise I get linking problems due to the additional functions).
And esp., if you don't like this approach, I need to redo my hands-on session.

File: doc/api.rst [hunk 1]
```diff
@@ -49,7 +49,7 @@ Three types of boundary condition are currently supported:
 Name             Description
 ===============  =========================================
 GO_BC_NONE       No boundary conditions are applied.
-GO_BC_EXTERNAL   Some external forcing is applied. This must be implemented by a kernel. The domain must be defined with a T-point mask (see :ref:`gocean1.0-grid-init`).
+GO_BC_EXTERNAL   Some external forcing is applied. This must be implemented by a kernel.
```
Comment (arporter) [inline doc/api.rst:52]: How come you've changed this? I would have thought that we still need a T-mask for the specification of where external BCs are applied.

File: finite_difference/src/grid_mod.f90 [hunk 2]
```diff
@@ -395,18 +393,6 @@ subroutine grid_init(grid, dxarg, dyarg, tmask)
...
        end do
-    else
-       ! No T-mask supplied. Check that grid has PBCs in both
-       ! x and y dimensions otherwise we won't know what to do.
-       if( .not. ( (grid%boundary_conditions(1) == GO_BC_PERIODIC) .and. &
-                   (grid%boundary_conditions(2) == GO_BC_PERIODIC) ) )then
-          call gocean_stop('grid_init: ERROR: No T-mask supplied and '// &
-                           'grid does not have periodic boundary conditions!')
-       end if
-       !> TODO add support for PBCs in parallel
```
Comment (arporter) [inline finite_difference/src/grid_mod.f90:406]: Does this mean we do now support PBCs in parallel? If not, we presumably still need to catch this situation?

File: doc/api.rst [hunk 3]
```diff
@@ -128,11 +125,17 @@ constructor::
...
 
  1. The grid on which the field exists
  2. The type of grid point at which the field is defined
     (``GO_U_POINTS``, ``GO_V_POINTS``, ``GO_T_POINTS`` or ``GO_F_POINTS``)
+ 3. ``do_tile``: If the field should be tiled among all threads, or if only
+    a single field should be allocated (which is not currently
+    supported by PSyclone).
+ 4. ``init_global_data``: a global 2D Fortran array, which must be
+    provided on each rank. On each rank the field will be initialised
+    with the data from the corresponding subdomain.
```
Comment (arporter) [inline doc/api.rst:153]: Please could you add a note here saying that the global array is simply to make life easier for a user (with small problem sizes) but may be omitted if this will cause problems with scaling to a large global domain size (and that it is then the application's responsibility to populate the field with the right data).

File: finite_difference/src/field_mod.f90 [hunk 4]
```diff
@@ -1287,6 +1305,79 @@ end function array_checksum
...
+    class(r2d_field), intent(in) :: self
+    real(go_wp), dimension(:,:),           &
+        allocatable, intent(out) :: global_data
+
+    real(go_wp), dimension(:), allocatable :: send_buffer, recv_buffer
+
+    integer :: dx, dy, ji, jj, i, n, rank, halo_x, halo_y
+    integer :: x_start, x_stop, y_start, y_stop
+
+    allocate(global_data(self%grid%global_nx, self%grid%global_ny))
```
Comment (arporter) [inline finite_difference/src/field_mod.f90]: Please could you check the error status of this allocate and abort if it fails.

File: finite_difference/src/field_mod.f90 [hunk 5]
```diff
@@ -1287,6 +1305,79 @@ end function array_checksum
...
+    endif
+
+    ! Determine maximum size of data to be sent. We don't need
+    ! to sent the halo, so reduce max_width and max_height by
+    ! 2*halo size.
+    halo_x = self%internal%xstart-1
+    halo_y = self%internal%ystart-1
+    n = (self%grid%decomp%max_width - 2*halo_x) *  &
+        (self%grid%decomp%max_height - 2*halo_y)
+    allocate(send_buffer(n))
```
Comment (arporter) [inline finite_difference/src/field_mod.f90]: Please could you also check the status of these two allocates.

**Review by arporter — CHANGES_REQUESTED** [review]
Apologies for forgetting about this Joerg. Generally I think it looks good although I have to admit it feels like a bit of a retrograde step to allow global arrays (I think we used to have them and then I removed them). However, since you've made them optional then I think it's OK as long as we're clear that the library can be used without them. (This is important if we want a simple benchmark capable of running with a very large domain.)
Please could you add a test of the new reduction functionality under `finite_difference/tests/dist_mem`. Apart from that, there's only a couple of minor things.

File: doc/api.rst (hunk 1 above)
Comment (hiker) [reply doc/api.rst:52]: The documentation in grid_mod.f90 states:
>  !! ... However, we also support the
     !! specification of a model consisting entirely of wet points

But that never worked, so I've fixed that.

Ah .... I just went back to the original commit, and realised it also had "... with PBC" :(  So, not really a fix as I thought. 
I would prefer not to require a tmask, it is really a small change (the rest works), leaving it up to the user how to setup the boundary condition. E.g. the test cases could all be simplified by removing the tmask code (it's just set to '1'), see my new test case.
Am happy to discuss if this should really be required (though it will make code easier if the need for tmask is just removed - to me feels the library is more generic this way).

File: doc/api.rst (hunk 3 above)
Comment (hiker) [reply doc/api.rst:153]: Done.

File: finite_difference/src/grid_mod.f90 (hunk 2 above)
Comment (hiker) [reply finite_difference/src/grid_mod.f90:406]: See comment above, but I have added the TODO back in. Removing this consistency check does not seem to cause any problems, the rest of the code just works.

File: finite_difference/src/field_mod.f90 (hunk 4 above)
Comment (hiker) [reply finite_difference/src/field_mod.f90]: Oops - done.

File: finite_difference/src/field_mod.f90 (hunk 5 above)
Comment (hiker) [reply finite_difference/src/field_mod.f90]: Oops again, and done.

**Comment by hiker** [comment]
Fixed, but am happy to discuss the requirement of the tmask in the telco tonight.

**Comment by hiker** [comment]
I grep'ed for t_mask in the source files, and there seems to be no other use of the tmask, except for copying it in the grid constructor (it is passed as a parameter somewhere, but then not actually used). So I don't see a good reason to insist on having a tmask, since all of the code should be working just fine if there is no tmask.

File: doc/api.rst (hunk 1 above)
Comment (arporter) [reply doc/api.rst:52]: I've double-checked the 'shallow' benchmark in PSycloneBench and it is fine...
With master of dl_esm_inf:
P initial CHECKSUM =   0.8192000000000004E+09
U initial CHECKSUM =   0.3258838709766664E+04
V initial CHECKSUM =   0.3258838709766665E+04
P CHECKSUM after   2000 steps =   0.8192000000000000E+09
U CHECKSUM after   2000 steps =   0.3258842842531393E+04
V CHECKSUM after   2000 steps =   0.3258842842531442E+04
With this branch:
psi initial CHECKSUM =   0.6641518585171037E+10
P initial CHECKSUM =   0.8192000000000004E+09
U initial CHECKSUM =   0.3258838709766664E+04
V initial CHECKSUM =   0.3258838709766665E+04
P CHECKSUM after   2000 steps =   0.8192000000000000E+09
U CHECKSUM after   2000 steps =   0.3258842842531393E+04
V CHECKSUM after   2000 steps =   0.3258842842531442E+04

However, I then realised that that is because the boundary-condition kernels in this app don't use the T-mask - they are hardwired to fill the first/last row/column of the domain.

Comment (arporter) [reply doc/api.rst:52]: MPI version of PSyclone-generated nemolite2d...
1 process:
ua checksum =   0.52670513E+02
va checksum =   0.57100321E+03
2 processes:
ua checksum =   0.52670513E+02
va checksum =   0.57100321E+03
4 processes:
ua checksum =   0.52670513E+02
va checksum =   0.57100321E+03
4 procs, with this version of dl_esm_inf:
ua checksum =   0.52670513E+02
va checksum =   0.57100321E+03

...so that's all good.

Comment (arporter) [reply doc/api.rst:52]: The point is that nemolite2d in PSycloneBench *does* use the T-mask to define an arbitrary domain. This is stored within the infrastructure so it can be supplied (in the PSy layer) to those kernels that require it, Therefore, I don't think we can remove it.
image

Comment (arporter) [reply doc/api.rst:52]: Therefore, it is really required. The reason PBCs aren't implemented is because we never got as far as 'PSyKAl'ing the 'shallow' benchmark which uses them - nemolite2d does not.

Comment (arporter) [reply doc/api.rst:52]: All in all, I think we need to revert this change!

File: doc/api.rst [hunk 6]
```diff
@@ -98,11 +98,8 @@ object. This is done via a call to the ``grid_init`` subroutine::
     !! wet (1), dry (0) or external (-1).
     integer, dimension(m,n), intent(in), optional :: tmask
 
-If no T-mask is supplied then this routine configures the grid
-appropriately for an all-wet domain with periodic boundary conditions
-in both the *x*- and *y*-dimensions. It should also be noted that
-currently only grids with constant resolution in *x* and *y* are
-supported by this routine.
+It should be noted that currently only grids with constant
```
Comment (arporter) [inline doc/api.rst]: Needs to be reverted? Or perhaps refined?

File: finite_difference/src/field_mod.f90 [hunk 7]
```diff
@@ -368,6 +373,17 @@ function r2d_field_constructor(grid,    &
     else
        self%data(:,:) = 0.0_go_wp
     end if
+
+    if (present(init_global_data)) then
+        dx = grid%subdomain%global%xstart - self%internal%xstart
+        dy = grid%subdomain%global%ystart - self%internal%ystart
+
```
Comment (arporter) [inline finite_difference/src/field_mod.f90:381]: This *could* be parallelised in the same way as the loop at L365 but since this is just initialisation it could be left for now, especially since the zeroing does first touch. Perhaps just a TODO here (no need for an issue number).

File: finite_difference/src/grid_mod.f90 [hunk 8]
```diff
@@ -67,9 +67,8 @@ module grid_mod
      !! This is the key quantity that determines the region that
      !! is actually simulated. However, we also support the
      !! specification of a model consisting entirely of wet points
-     !! with periodic boundary conditions. Since this does not
-     !! require a T-mask, we do not allocate this array for that
-     !! case.
+     !! Since this does not require a T-mask, we do not allocate
```
Comment (arporter) [inline finite_difference/src/grid_mod.f90]: Please revert/refine in light of the fact that one day we *might* support PBCs.

File: finite_difference/src/grid_mod.f90 [hunk 9]
```diff
@@ -294,8 +293,7 @@ end function grid_constructor
   !! @param[in] dxarg Grid spacing in x dimension
   !! @param[in] dyarg Grid spacing in y dimension
   !! @param[in] tmask Array holding the T-point mask which defines
-  !!                  the contents of the local domain. Need not be
-  !!                  supplied if domain is all wet and has PBCs.
+  !!                  the contents of the local domain.
```
Comment (arporter) [inline finite_difference/src/grid_mod.f90]: Please revert.

File: finite_difference/src/grid_mod.f90 (hunk 2 above)
Comment (arporter) [reply finite_difference/src/grid_mod.f90:406]: I'd still like to keep the consistency check - an error message is always helpful :-)

File: finite_difference/tests/dist_mem/test_reduction.f90 [hunk 10]
```diff
@@ -0,0 +1,195 @@
+!------------------------------------------------------------------------------
+! BSD 2-Clause License
+! 
+! Copyright (c) 2019, Science and Technology Facilities Council
```
Comment (arporter) [inline finite_difference/tests/dist_mem/test_reduction.f90]: Please update the date.

File: finite_difference/tests/dist_mem/test_reduction.f90 [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+!------------------------------------------------------------------------------
+! Author: J. Henrichs, Bureau of Meteorology
+! Author: A. R. Porter, STFC Daresbury Laboratory
+
+!> Tests for the halo-exchange functionality of the dl_esm_inf library.
```
Comment (arporter) [inline finite_difference/tests/dist_mem/test_reduction.f90]: s/halo-exchange/reduction/

File: finite_difference/tests/dist_mem/test_reduction.f90 [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    integer :: ji, jj
+    real(go_wp) :: global_value
+    
+    do jj = 1, size(global_field, 2)
+       do ji = 1, size(global_field, 1)
+         global_value = unique_global_value(ji, jj, size(global_field, 1)) + 1
+         if (global_field(ji, jj) /= global_value) then
+            print *, "Incorrect global field data gathered for ", ji, jj, &
+                     global_field(ji, jj), global_value
+            !call gocean_stop('Failed to gather local data')
```
Comment (arporter) [inline finite_difference/tests/dist_mem/test_reduction.f90]: Please remove this line.

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for adding the tests - they're nice :-)
After much thought and examination of PSycloneBench, I've concluded that we do need the T-mask so have requested you revert some of those changes.
If you agree then once that's done this is ready to go.

File: finite_difference/tests/dist_mem/test_reduction.f90 (hunk 12 above)
Comment (hiker) [reply finite_difference/tests/dist_mem/test_reduction.f90]: Done.

File: finite_difference/tests/dist_mem/test_reduction.f90 (hunk 11 above)
Comment (hiker) [reply finite_difference/tests/dist_mem/test_reduction.f90]: Done.

File: finite_difference/tests/dist_mem/test_reduction.f90 (hunk 10 above)
Comment (hiker) [reply finite_difference/tests/dist_mem/test_reduction.f90]: Oops.

**Comment by hiker** [comment]
> [finite_difference/src/grid_mod.f90](https://github.com/stfc/dl_esm_inf/pull/65/files/483a0a31b77f9e22fa0e323bac796a97a21412bc#diff-b2c203ffc01de9f6b7b751d80e68d892e2c019f04f72b9dc14465e274643da26)
          call gocean_stop('grid_init: ERROR: No T-mask supplied and '// &
                           'grid does not have periodic boundary conditions!')
       end if
       !> TODO add support for PBCs in parallel
Member
@[arporter](https://github.com/arporter) arporter [3 days ago](https://github.com/stfc/dl_esm_inf/pull/65#discussion_r1467451427)
I'd still like to keep the consistency check - an error message is always helpful :-)

Sorry, I don't know how to quote-reply to that comment, I don't have the usual 'reply' option.

As far as I can see, the state if we don't have a T-mask is either:
1. We have periodic boundary conditions. Then indeed you don't need to supply a tmask ... but MPI execution aborts later(!) at a halo exchange since this is not yet supported.
2. We don't have periodic boundary conditions. The above test will then abort the execution as well.

Given that 2. actually works, shouldn't it be the other way round? We abort early if we detect an unsupported configuration (instead of running and aborting later)?

I'll try to have a chat with you about this tonight. I feel we might have a misunderstanding. My change only makes the T-mask truly optional, all existing code that uses a T-mask will continue to work, but in addition we can also use it without T-mask, which we can't atm. The only configuration that would be a problem is if an application does not provide a tmask, but in a kernel then asks for it (though we could detect this and then either provide with an error, or even provide a dummy tmask, i.e. same size as the grid, filled with 1's)

File: finite_difference/src/field_mod.f90 (hunk 7 above)
Comment (hiker) [reply finite_difference/src/field_mod.f90:381]: Done - I mean I added omp :)

**Comment by hiker** [comment]
Hmm - I found a comment 'Please update the date.' for grid_mod.f90 ... since it had no license, I've added the usual license to the top of the file. But when trying to comment on that in the review here, it turns out that comment was for field_mod (which I had already fixed). Given that it seems correct to add a proper license for grid_mod, I left it in.

File: doc/api.rst (hunk 1 above)
Comment (hiker) [reply doc/api.rst:52]: In our telco we agreed that we should modify the test for PBC to abort if we are in an unsupported configutation if no tmask is specified (i.e. abort in distributed memory, but allow single process, since this is what e.g. the shallow benchmark uses).

If no tmask is specified, we add a dummy tmask (set to all wet) based on the specified grid dimension. This way any kernel that requests a tmask will get valid data.

I've also reworked the documentation to reflect this.

File: finite_difference/src/grid_mod.f90 (hunk 2 above)
Comment (hiker) [reply finite_difference/src/grid_mod.f90:406]: As agreed, the test will now abort in an unsupported configuration (PCB with no tmask in case of distributed memory runs).

File: doc/api.rst (hunk 6 above)
Comment (hiker) [reply doc/api.rst]: I've updated the documentation to reflect what we agreed on (I think :)  )

File: finite_difference/src/grid_mod.f90 (hunk 8 above)
Comment (hiker) [reply finite_difference/src/grid_mod.f90]: Docs updated.

File: finite_difference/src/grid_mod.f90 (hunk 9 above)
Comment (hiker) [reply finite_difference/src/grid_mod.f90]: Docs updated.

**Comment by hiker** [comment]
I think I have updated everything as discussed. Back to you :)

File: finite_difference/src/grid_mod.f90 (hunk 2 above)
Comment (arporter) [reply finite_difference/src/grid_mod.f90:406]: Thanks :-)

**Review by arporter — APPROVED** [review]
All good now - thanks.
I've checked the `make_gen` target of PSycloneBench with this branch and it was fine.
The tests added in this PR also work fine.
Will proceed to merge.
